internal/encryption/method/aesgcm/README.md

```diff
@@ -0,0 +1,49 @@
+# AES-GCM encryption method
+
+This folder contains the state encryption implementation of the AES-GCM encryption method ([NIST SP 800-38D](https://csrc.nist.gov/pubs/sp/800/38/d/final)).
+
+## Configuration
+
+You can configure the encryption by specifying the following method block:
+
+```hcl2
+terraform {
+  encryption {
+    method "aes_gcm" "mymethod" {
+      key        = "key here" # Pass your 16, 24, or 32 byte encryption key here.
+      aad        = ""         # Leave empty unless needed
+      nonce_size = 12         # Minimum: 1, do not change unless needed
+      tag_size   = 16         # Valid values: 12-16, do not change unless needed
+    }
+  }
+}
+```
+
+| Field              | Default | Description                                                                                                                                                                                      |
+|--------------------|---------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `key` (*required*) |         | Your encryption key in binary format, likely referencing a key provider. Must be 16, 24, or 32 bytes long.                                                                                       |
+| `aad`              |         | Additional Authenticated Data. This data is stored along the encrypted form and authenticated. The AAD value of the encrypted form must match the configuration, otherwise the decryption fails. |
+| `nonce_size`       | 12      | Size of the Initialization Vector for AES-GCM. This field should not be changed and is provided for future compatibility.                                                                        |
+| `tag_size`         | 16      | Size of the authentication tag (12-16 bytes). This field should not be changed and is provided for future compatibility.                                                                         |
+
+## Key rotation
+
+AES-GCM keys have a limited lifetime of `2^32` blocks, equaling roughly 64 GB of data that can be encrypted before the keys should be considered compromised. Users should rotate keys well before this limit is reached.

```

Comment:
Go's default, and NIST's recommendation, is that safe balance. There is no better value than 12 bytes given a 32-bit counter.

I think [NIST SP800-38D](https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-38d.pdf) Section 8.3 Constraints on the Number of Invocations is the guidance you're looking for.

Practically though, you don't get any benefit in changing your nonce size with Go; either way you're still limited to an (implicit) 2^32 counter size (which 16-byte nonces don't really help: now you can have two nonces `A`, `B` with `A != B` such that `A+j == B+K` and you break the cryptosystem, just in a more convoluted way. A least with 12-byte nonces you can ensure that as long as `A` & `B` are distinct, you'll never have `A+j == B+k`).

All this to say, you _do_ have to worry about rotation -- and you're not getting out of it with AES-GCM.

Reply:
Our current understanding is that if we use key-derivation functions, such as PBKDF2 or Argon2 in conjunction with AES-GCM, this is no longer a problem. Is this correct @cipherboy ?

Reply:
Now, I will say, some things sometimes called PBKDFs aren't suitable for _key_ derivation and only for password storage, e.g., bcrypt and scrypt incorporate entropy  (randomness) into each password hash and thus generate different "keys"/outputs each time and thus won't work for this. You need a deterministic scheme.

Reply:
> Every time a user (whether intentionally or unintentionally, e.g., everyone using password1234) uses the same password for encryption, they get the same AES key as output and thus runs into this problem. They'd then have to rotate passwords, or better, choose a strong enough password that roughly looks like a key and use it only once. However, now you don't know if someone else choose the same password, whereas you could be guaranteed nobody else chose the same 128-to-256-bit AES key. :-)

If I understand correctly, you'd be generating a new salt every time you're generating an encryption key, which means a new encryption key whenever you're encrypting.

Reply:
There's some notes about it here: https://crypto.stackexchange.com/questions/84439/is-it-dangerous-to-encrypt-lots-of-small-files-with-the-same-key

But basically, both pbkdf2 and argon2 accept a salt. If we use an 8-byte or 16-byte salt, then I believe we're safe, and a passphrase can be reused without worrying about key reuse.

Reply:
(we're planning to store the salt as part of the statefile metadata field, which is stored unencrypted)

Reply:
Ah, ok, thanks!

This still ends up being a trade off: you trade a known 128-to-256-bits of security for a questionable security margin in the password itself. Password management and injection is no easier or harder than key management, and you (usually) cannot take advantage of things like HSM storage.

I don't think its a silver bullet, even with public salt, as any weak passwords can still be cracked relatively easily. It is why things like PKCS#12 aren't widely adopted outside of niche communities/use cases and why unencrypted PEM ultimately prevails. If you're just putting the password along side the file itself (so a service can read it), its not different than having an unencrypted blob, IMO.

---

Why have externally specified keys outside of a PKCS#11 interface? Why not have OpenTofu generate a random key on encryption (when migrating from unencrypted->encrypted; otherwise reuse the same specified key, perhaps with an internal counter) and have the user manage that as a password (a la encrypted backups of things like Signal or Matrix)?

My 2c., but I suppose there's no easy design here, otherwise it would've already been done :-)

Reply:
> This still ends up being a trade off: you trade a known 128-to-256-bits of security for a questionable security margin in the password itself. Password management and injection is no easier or harder than key management, and you (usually) cannot take advantage of things like HSM storage.

However, assuming a good 32-byte password with a (randomly generated on each encryption) 16-byte salt, you do at least get protection from reusing IV's for an AES key in GCM, unless I'm missing something.

Of course, a user could still just write a script to generate a new encryption key, store it in a key management system, fetch the previous one, and then pass them as the encryption/decryption keys to opentofu, every time it's ran.

In practice, we do expect most users to utilize key providers like AWS KMS, which mostly handle everything for you with their managed keys.

Reply:
You could get that same protection with a regular KDF (HKDF or an SP 800-108 KBKDF) with published salt though. :-)

The goal of adding a PB KDF shouldn't be to avoid nonce reuse. It should be because it's a workflow you want to encourage (use a password to unlock/share X).

Reply:
Let me know if you want me to build a PKCS#11 compatible SP800-108 KDF library. I've written them a few times and I wouldn't mind trying to upstream it to x/crypto, though I note they've not taken many PRs from me that quickly :-D

Reply:
Thanks for that.

That said, I suppose there's no big reason **not** to use pbkdf2 here though, right? Mainly, I see that hkdf, pbkdf2 and argon2 **are** in the golang.org/x/crypto package, while the only two Go implementations of `SP800-108 Feedback mode KDF` I found are 1. [an unmaintained repo with one star](https://pkg.go.dev/github.com/chrisccoulson/go-sp800.108-kdf) 2. [HashiCorp's implementation in the Vault sdk](https://pkg.go.dev/github.com/hashicorp/vault/sdk/helper/kdf) (which, fwiw, is still MPL lincensed).

But I agree with you that we might want to revise the messaging here. I don't think we expect or encourage people to use memorable passwords, and would rather encourage generating a long password and keeping it in something like aws secrets manager, or 1password.

Reply:
That is a counter mode KDF, I was thinking of a pipeline mode KDF as that mirrors HKDF. PKCS#11 and SP800-108 has additional flexibility that the linked code doesn't (also present in OpenBao fwiw) for parameter ordering that let's you fenangle pipeline mode to behave exactly as the second stage of HKDF (Expand).

Reply:
I wonder if you could get away with the opposite argument, tbh, and just call HKDF_Expand and pretend it is SP800-108 Feedback if anyone asks. :D

Reply:
Just to clarify, is there any reason I'm not seeing (other than users possibly getting the wrong message about expected passphrase complexity, which we can handle via documentation) not to just go with pbkdf2 here?

Reply:
Nah, if it's the UX you want, go for it. But if you're passing in "real" keys to it, I think NIST would frown on that for FIPS.

Reply:
That makes sense, we can just recommend sufficiently complex "passphrases".

Thanks a lot for this discussion @cipherboy, I really appreciate you joining the discussion here!

Reply:
No problem, glad the feedback was useful!